talisker/metrics.py

```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
             client = talisker.statsd.get_client()
             name = self.get_statsd_name(labels)
             client.incr(name, amount)
+
+
+histogram_archive = 'histogram_archive.db'
+counter_archive = 'counter_archive.db'
+
+
+def histogram_sorter(sample):
+    # sort histogram samples in order of bucket size
+    name, labels, _ = sample
+    return name, float(labels.get('le', 0))
+
+
+def prometheus_cleanup_worker(pid):
+    """Clean up after a multiprocess worker has died."""
+    if prometheus_client is None:
+        return
+
+    try:
+        # just deletes delete gauge files
+        multiprocess.mark_process_dead(pid)
+        prom_dir = os.environ['prometheus_multiproc_dir']
+        worker_files = [
+            'histogram_{}.db'.format(pid),
+            'counter_{}.db'.format(pid),
+        ]
+        paths = [os.path.join(prom_dir, f) for f in worker_files]
+
+        # check at least one worker file exists
+        if not any(os.path.exists(path) for path in paths):
+            return
+
+        files = [
+            os.path.join(prom_dir, f) for f in
+            [histogram_archive, counter_archive] + worker_files
+        ]
+        metrics = collect(files)
+
+        tmp_histogram = tempfile.NamedTemporaryFile(delete=False)
+        tmp_counter = tempfile.NamedTemporaryFile(delete=False)
+        write_metrics(metrics, tmp_histogram.name, tmp_counter.name)
+
+        os.rename(
+            tmp_histogram.name, os.path.join(prom_dir, histogram_archive))
+        os.rename(
+            tmp_counter.name, os.path.join(prom_dir, counter_archive))
+
+        for path in paths:
+            if os.path.exists(path):
+                os.unlink(path)
+
+    except Exception:
+        # we should never fail at cleaning up
```

Comment:
agree with the sentiment, let's put it in the caller though? i.e. `prometheus_multiprocess_worker_exit`